lib/ts/MemArena.h

```diff
@@ -0,0 +1,134 @@
+    with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+ */
+
+#ifndef _MEM_ARENA_H_
+#define _MEM_ARENA_H_
+
+#include <mutex>
+#include <ts/MemSpan.h>
+
+/// Apache Traffic Server commons.
+namespace ts
+{
+/** MemArena is a memory arena for allocations.
+
+    The intended use is for allocating many small chunks of memory - few, large allocations are best handled independently.
+    The purpose is to amortize the cost of allocation of each chunk across larger allocations in a heap style. In addition the
+    allocated memory is presumed to have similar lifetimes so that all of the memory in the arena can be de-allocatred en masse.
+
+    A generation is essentially a block of memory. The normal workflow is to freeze() the current generation, alloc() a larger and
+    newer generation, copy the contents of the previous generation to the new generation, and then thaw() the previous generation.
+    Note that coalescence must be done by the caller because MemSpan will only give a reference to the underlying memory.
+ */
+class MemArena
+{
+public:
+  /** Simple internal arena block of memory. Maintains the underlying memory.
+   */
+  struct MemArenaBlock {
+    void *data;
+    size_t size;
+    size_t allocated;
+    MemArenaBlock *next;
+
+    void alloc(size_t n);
+    void free(); ///< Need to free() underlying memory before delete.
+  };
+
+  MemArena();
+  MemArena(size_t n);
+
+  ~MemArena();
+
+  /** MemSpan alloc(size_t n)
+
+      Returns a span of memory within the arena. alloc() is self expanding but DOES NOT self coalesce. This means
+      that no matter the arena size, the caller will always be able to alloc() @a n bytes.
+
+      @param n number of bytes to allocate.
+      @return a MemSpan of the allocated memory.
+      @note in order to guarantee n bytes of memory, the implementation keeps creating new
+              generations until there is a large enough memory chunk. If you need a large chunk of memory,
+              you should be creating a large MemArena and then allocating from it. Rather than relying on the
+              default size and arena resizing itself.
+   */
+  MemSpan alloc(size_t n);
+
+  /** MemArena& freeze(size_t n = 0)
+
+      Will "freeze" a generation of memory. Any memory previously allocated can still be used. This is an
+      important distinction as freeze does not mean that the memory is immutable, only that subsequent allocations
+      will be in a new generation.
+
+      @param n Number of bytes for new generation.
+        if @a n == 0. double in size.
```

Comment:
No. If not specified the new block should be large enough to hold all existing allocations.